File: String/1392. Longest Happy Prefix.cpp
/**
 * Rolling Hash
 * 
 */

class Solution {
public:
    string longestPrefix(string &s) {
        long h1 = 0, h2 = 0, mul = 1, len = 0, mod = 1000000007;
        for (int i = 0, j = s.length() - 1; j > 0; ++i, --j) {
            int first = s[i] - 'a', last = s[j] - 'a';
            h1 = (h1 * 26 + first) % mod;
            h2 = (h2 + mul * last) % mod;
            mul = mul * 26 % mod;
            if (h1 == h2) { // Collision-Aware Incremental Hash
            if (s.compare(0, i + 1 - len, s, j, s.size() - j - len) == 0 &&
               s.compare(len, i + 1 - len, s, j + len, s.size() - j - len) == 0)
                len = i + 1;
            }
        }
        return s.substr(0, len);
    }
};

/**
 * KMP
 */

class Solution {
public:
    string longestPrefix(string &s) {
        vector<int> dp(s.size());
        int j = 0;
        for (int i = 1; i < s.size(); ++i) {
            if (s[i] == s[j])
                dp[i] = ++j;
            else if (j > 0) {
                j = dp[j - 1];
                --i;
            }
        }
        return s.substr(0, j);
    }
};

File: String/1392. Longest Happy Prefix.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        k = [0]
        j = 0 
        for i in s[1:]:
            while j > 0 and i != s[j]:
                j = k[j-1]
            if i == s[j]:
                j += 1 
            k.append(j)
        return s[:k[-1]]

"""
Rolling Hash
"""

class Solution:
    def longestPrefix(self, s):
            # res stores the index of the end of the prefix, used for output the result
        # l stores the hash key for prefix
        # r stores the hash key for suffix
        # mod is used to make sure that the hash value doesn't get too big, you can choose another mod value if you want.
        res, l, r, mod = 0, 0, 0, 10**9 + 7

        # now we start from the beginning and the end of the string
        # note you shouldn't search the whole string! because the longest prefix and suffix is the string itself
        for i in range(len(s) - 1):

            # based on an idea that is similar to prefix sum, we calculate the prefix hash in O(1) time.
            # specifically, we multiply the current prefix by 128 (which is the length of ASCII, but you can use another value as well)
            # then add in the ASCII value of the upcoming letter
            l = (l * 128 + ord(s[i])) % mod

            # similarly, we can calculate the suffix hash in O(1) time.
            # Specifically, we get the ith letter from the end using s[~i], note ~i is -i-1
            # we find the pow(128, i, mod) and multiply by the letter's ASCII value
            # Actually, if we don't care about the beautifulness of the code, you can have a variable to keep track of pow(128, i, mod) as you increase i
            r = (r + pow(128, i, mod) * ord(s[~i])) % mod

           # we check if the prefix and suffix agrees, if yes, we find yet another longer prefix, so we record the index
            if l == r: res = i + 1

       # after we finish searching the string, output the prefix
        return s[:res]